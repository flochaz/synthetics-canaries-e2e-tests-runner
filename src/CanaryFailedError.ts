export default class CanaryFailedError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'CanaryFailedError';
  }
}