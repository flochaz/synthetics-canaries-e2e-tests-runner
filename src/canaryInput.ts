import * as cdk from '@aws-cdk/core';

export interface ICanaryInput {
  name: string;
  value: cdk.CfnOutput;
}