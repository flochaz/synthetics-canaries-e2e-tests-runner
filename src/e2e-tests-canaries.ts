import * as synthetics from '@aws-cdk/aws-synthetics';
import * as cdk from '@aws-cdk/core';

export interface E2ETestsCanariesProps {
  endpointUnderTest: string;
}

export class E2ETestsCanaries extends cdk.Construct {
  public readonly canaries: synthetics.Canary[] = [];
  constructor(scope: cdk.Construct, id: string, props: E2ETestsCanariesProps) {
    super(scope, id);


    this.canaries.push(new synthetics.Canary(this, 'Test_1', {
      canaryName: 'test-1',
      schedule: synthetics.Schedule.once(),
      test: synthetics.Test.custom({
        code: synthetics.Code.fromInline('exports.handler = function(event) { console.log(`Will test ${process.env.ENDPOINT_UNDER_TEST}`); }'),
        handler: 'index.handler',
      }),
      runtime: synthetics.Runtime.SYNTHETICS_NODEJS_PUPPETEER_3_1,
      environmentVariables: {
        ENDPOINT_UNDER_TEST: props.endpointUnderTest,
      },
    }));

    this.canaries.push(new synthetics.Canary(this, 'Test_2', {
      canaryName: 'test_2',
      schedule: synthetics.Schedule.once(),
      test: synthetics.Test.custom({
        code: synthetics.Code.fromInline('exports.handler = function(event, ctx, cb) { return cb(null, "hi 2"); }'),
        handler: 'index.handler',
      }),
      runtime: synthetics.Runtime.SYNTHETICS_NODEJS_PUPPETEER_3_1,
      environmentVariables: {
        ENDPOINT_UNDER_TEST: props.endpointUnderTest,
      },
    }));

  }
}