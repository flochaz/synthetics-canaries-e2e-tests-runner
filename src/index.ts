export * from './orchestrator';
export * from './e2e-test-step';
export * from './canaryInput';