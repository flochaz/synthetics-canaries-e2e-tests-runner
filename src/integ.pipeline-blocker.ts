import * as synthetics from '@aws-cdk/aws-synthetics';
import * as cdk from '@aws-cdk/core';
import * as cdkpipeline from '@aws-cdk/pipelines';
// import E2ETestsStep from './e2e-test-step';

const mockApp = new cdk.App();


/**
 * Stack to hold the pipeline
 */
class MyPipelineStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const pipeline = new cdkpipeline.CodePipeline(this, 'Pipeline', {
      synth: new cdkpipeline.ShellStep('Synth', {
        // Use a connection created using the AWS console to authenticate to GitHub
        // Other sources are available.
        input: cdkpipeline.CodePipelineSource.connection('flochaz/synthetics-canaries-e2e-tests-runner', 'test', {
          connectionArn: 'arn:aws:codestar-connections:eu-west-1:036129959679:connection/c92f72f6-1838-49f7-be70-3cd461dbb227', // Created using the AWS console * });',
        }),
        commands: [
          'npx projen build',
          'cdk synth --app=./lib/integ.pipeline-blocker.js',
        ],
      }),
    });

    // 'MyApplication' is defined below. Call `addStage` as many times as
    // necessary with any account and region (may be different from the
    // pipeline's).
    const myApp = new MyApplication(this, 'Demo', {});

    // const e2eTestsRunnerStep = new E2ETestsStep(this, myApp.canaries, 'E2ETestsRunner');

    pipeline.addStage(myApp);
  }
}

/**
 * Your application
 *
 * May consist of one or more Stacks (here, two)
 *
 * By declaring our DatabaseStack and our ComputeStack inside a Stage,
 * we make sure they are deployed together, or not at all.
 */
class MyApplication extends cdk.Stage {
  public readonly canaries: synthetics.Canary[] = [];

  constructor(scope: cdk.Construct, id: string, props?: cdk.StageProps) {
    super(scope, id, props);

    const e2eTestsStack = new cdk.Stack(this, 'testing-stack');


    this.canaries.push(new synthetics.Canary(e2eTestsStack, 'Test_1', {
      canaryName: 'test-1',
      schedule: synthetics.Schedule.once(),
      test: synthetics.Test.custom({
        code: synthetics.Code.fromInline('exports.handler = function(event) { console.log("hi 1"); }'),
        handler: 'index.handler',
      }),
      runtime: synthetics.Runtime.SYNTHETICS_NODEJS_PUPPETEER_3_1,
    }));

    this.canaries.push(new synthetics.Canary(e2eTestsStack, 'Test_2', {
      canaryName: 'test_2',
      schedule: synthetics.Schedule.once(),
      test: synthetics.Test.custom({
        code: synthetics.Code.fromInline('exports.handler = function(event, ctx, cb) { return cb(null, "hi 2"); }'),
        handler: 'index.handler',
      }),
      runtime: synthetics.Runtime.SYNTHETICS_NODEJS_PUPPETEER_3_1,
    }));
  }
}

// In your main file
new MyPipelineStack(mockApp, 'PipelineStack', {});