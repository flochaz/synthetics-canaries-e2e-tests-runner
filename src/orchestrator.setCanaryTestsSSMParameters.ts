import * as AWS from 'aws-sdk';
const ssmClient = new AWS.SSM();


exports.handler = async (event: any) => {
  if (event.inputVariables) {
    console.log(`storing canary input parameters: ${JSON.stringify(event)}`);
    for (const inputVar of event.inputVariables) {
      await ssmClient.putParameter({
        Name: inputVar.name,
        Value: inputVar.value,
        Type: 'String',
        Overwrite: true,
      }).promise();
    }
  } else {
    console.log('No input provided. skipping.');
  }
  return;
};
