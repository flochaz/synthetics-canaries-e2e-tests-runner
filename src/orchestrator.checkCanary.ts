import * as synthetics from '@aws-cdk/aws-synthetics';
import * as AWS from 'aws-sdk';
import CanaryFailedError from './CanaryFailedError';

const synthteticsClient = new AWS.Synthetics();


exports.handler = async (event: synthetics.Canary): Promise<AWS.Synthetics.CanaryRunStatus> => {

  let result: AWS.Synthetics.CanaryRunStatus =
  { State: 'RUNNING' };
  try {
    console.log(`check canary ${event.canaryName}`);
    const runs = await synthteticsClient.getCanaryRuns({
      Name: event.canaryName,
      MaxResults: 1,
    }).promise();
    if (!runs.CanaryRuns || runs.CanaryRuns.length === 0) {
      console.error(`No run found for canary ${event.canaryName}`);
      throw new Error(`No run found for canary ${event.canaryName}`);
    }
    console.log(`result: ${JSON.stringify(runs.CanaryRuns[0])}`);
    result = runs.CanaryRuns[0]!.Status!;
    if (result.State === 'FAILED') {
      throw new CanaryFailedError(result.StateReason!);
    }
  } catch (error) {
    console.error(`Something went wrong checking ${event.canaryName}: ${error}`);
    throw new Error(error);
  }
  return result;
};

