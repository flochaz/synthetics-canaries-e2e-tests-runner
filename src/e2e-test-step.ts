import * as codepipeline from '@aws-cdk/aws-codepipeline';
import * as codepipeline_actions from '@aws-cdk/aws-codepipeline-actions';
import * as sfn from '@aws-cdk/aws-stepfunctions';
import * as synthetics from '@aws-cdk/aws-synthetics';
import * as cdk from '@aws-cdk/core';
import * as cdkpipeline from '@aws-cdk/pipelines';
import { ICanaryInput } from './canaryInput';
import { StepFunctionOrchestrator } from './orchestrator';

/**
 * Construction properties for a `E2ETestsStep`.
 */
export interface E2ETestsStepProps {

  /**
   * Array of AWS Cloudwatch canary to execute in this step.
   */
  readonly canaries: synthetics.Canary[];

  /**
   * Scope in wich to instantiate the state machine (usually your pipeline stack).
   */
  readonly scope: cdk.Construct;

  /**
   * The potential list of CloudFormation outputs exposed by
   * the App under test deployed in the previous step of the code pipeline workflow
   * and that are needed by canaries to run properly.
   * Those will be pushed to AWS SSM Parameter store to be accessed by the canary at runtime
   *  using the name set in the ICanaryInput.
   */
  readonly inputsFromDeployedStack: ICanaryInput[];
}

/**
 * Run AWS Cloudwatch Canaries end to end tests in parallel in the pipeline
 */
export class E2ETestsStep extends cdkpipeline.Step implements cdkpipeline.ICodePipelineActionFactory {
  public readonly stateMachine: sfn.StateMachine;
  public readonly inputsFromDeployedStack: any[] = [];
  constructor(id: string, props: E2ETestsStepProps) {
    super(id);

    const e2eTestsRunner = new StepFunctionOrchestrator(props.scope, 'E2ETEstsRunner', {
      canaries: props.canaries,
    });

    this.stateMachine = e2eTestsRunner.stateMachine;
    for (const input of props.inputsFromDeployedStack) {
      const namespace = cdk.Stack.of(input.value).artifactId;
      const variableName = input.value.exportName;
      this.inputsFromDeployedStack.push({ name: input.name, value: `#{${namespace}.${variableName}}` });
    }
  }

  public produceAction(stage: codepipeline.IStage, options: cdkpipeline.ProduceActionOptions): cdkpipeline.CodePipelineActionFactoryResult {

    stage.addAction(new codepipeline_actions.StepFunctionInvokeAction({
      stateMachine: this.stateMachine,
      runOrder: options.runOrder,
      actionName: options.actionName,
      stateMachineInput: codepipeline_actions.StateMachineInput.literal(this.inputsFromDeployedStack),
    }),
    );

    return { runOrdersConsumed: 1 };
  }
}
