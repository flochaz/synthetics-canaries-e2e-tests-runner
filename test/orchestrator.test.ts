import { SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import * as synthetics from '@aws-cdk/aws-synthetics';
import * as cdk from '@aws-cdk/core';
import { StepFunctionOrchestrator } from '../src';
test('minimal usage', () => {
  // GIVEN
  const stack = new cdk.Stack();

  const canary = new synthetics.Canary(stack, 'MyCanary', {
    schedule: synthetics.Schedule.rate(cdk.Duration.minutes(5)),
    test: synthetics.Test.custom({
      code: synthetics.Code.fromInline('console.log("Hello World!");'),
      handler: 'index.handler',
    }),
    runtime: synthetics.Runtime.SYNTHETICS_NODEJS_PUPPETEER_3_1,
  });


  // WHEN
  new StepFunctionOrchestrator(stack, 'stepFunctionOrchestrator', {
    canaries: [canary],
  });

  // THEN
  expect(SynthUtils.synthesize(stack).template).toMatchSnapshot();
});