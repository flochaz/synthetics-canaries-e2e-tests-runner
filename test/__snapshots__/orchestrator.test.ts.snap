// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`minimal usage 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters49cd87fb06607303bf223aa06c1cef28acff95117850c08947a62d8257902d6dArtifactHashE7766949": Object {
      "Description": "Artifact hash for asset \\"49cd87fb06607303bf223aa06c1cef28acff95117850c08947a62d8257902d6d\\"",
      "Type": "String",
    },
    "AssetParameters49cd87fb06607303bf223aa06c1cef28acff95117850c08947a62d8257902d6dS3Bucket81F354A2": Object {
      "Description": "S3 bucket for asset \\"49cd87fb06607303bf223aa06c1cef28acff95117850c08947a62d8257902d6d\\"",
      "Type": "String",
    },
    "AssetParameters49cd87fb06607303bf223aa06c1cef28acff95117850c08947a62d8257902d6dS3VersionKey17FB176D": Object {
      "Description": "S3 key for asset version \\"49cd87fb06607303bf223aa06c1cef28acff95117850c08947a62d8257902d6d\\"",
      "Type": "String",
    },
    "AssetParametersa092c9a65e948c08fdd58344d465f3fe147850fd89e1402f7354ce067360eb03ArtifactHash87FFE166": Object {
      "Description": "Artifact hash for asset \\"a092c9a65e948c08fdd58344d465f3fe147850fd89e1402f7354ce067360eb03\\"",
      "Type": "String",
    },
    "AssetParametersa092c9a65e948c08fdd58344d465f3fe147850fd89e1402f7354ce067360eb03S3Bucket7E06B8E8": Object {
      "Description": "S3 bucket for asset \\"a092c9a65e948c08fdd58344d465f3fe147850fd89e1402f7354ce067360eb03\\"",
      "Type": "String",
    },
    "AssetParametersa092c9a65e948c08fdd58344d465f3fe147850fd89e1402f7354ce067360eb03S3VersionKey8A44DBB6": Object {
      "Description": "S3 key for asset version \\"a092c9a65e948c08fdd58344d465f3fe147850fd89e1402f7354ce067360eb03\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "MyCanary1A94CAFA": Object {
      "Properties": Object {
        "ArtifactS3Location": Object {
          "Fn::Join": Array [
            "",
            Array [
              "s3://",
              Object {
                "Ref": "MyCanaryArtifactsBucket89975E6D",
              },
            ],
          ],
        },
        "Code": Object {
          "Handler": "index.handler",
          "Script": "console.log(\\"Hello World!\\");",
        },
        "ExecutionRoleArn": Object {
          "Fn::GetAtt": Array [
            "MyCanaryServiceRole593F9DD9",
            "Arn",
          ],
        },
        "Name": "mycanary",
        "RuntimeVersion": "syn-nodejs-puppeteer-3.1",
        "Schedule": Object {
          "DurationInSeconds": "0",
          "Expression": "rate(0 minute)",
        },
        "StartCanaryAfterCreation": true,
      },
      "Type": "AWS::Synthetics::Canary",
    },
    "MyCanaryArtifactsBucket89975E6D": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "MyCanaryServiceRole593F9DD9": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "s3:ListAllMyBuckets",
                  "Effect": "Allow",
                  "Resource": "*",
                },
                Object {
                  "Action": Array [
                    "s3:PutObject",
                    "s3:GetBucketLocation",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        Object {
                          "Fn::GetAtt": Array [
                            "MyCanaryArtifactsBucket89975E6D",
                            "Arn",
                          ],
                        },
                        "/*",
                      ],
                    ],
                  },
                },
                Object {
                  "Action": "cloudwatch:PutMetricData",
                  "Condition": Object {
                    "StringEquals": Object {
                      "cloudwatch:namespace": "CloudWatchSynthetics",
                    },
                  },
                  "Effect": "Allow",
                  "Resource": "*",
                },
                Object {
                  "Action": Array [
                    "logs:CreateLogStream",
                    "logs:CreateLogGroup",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":logs:::*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "canaryPolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "stepFunctionOrchestratorE2ETestsRunner417F497C": Object {
      "DependsOn": Array [
        "stepFunctionOrchestratorE2ETestsRunnerRoleDefaultPolicy267BF1ED",
        "stepFunctionOrchestratorE2ETestsRunnerRole6F118857",
      ],
      "Properties": Object {
        "DefinitionString": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"StartAt\\":\\"parallel\\",\\"States\\":{\\"parallel\\":{\\"Type\\":\\"Parallel\\",\\"Next\\":\\"checkResults\\",\\"Branches\\":[{\\"StartAt\\":\\"",
              Object {
                "Ref": "MyCanary1A94CAFA",
              },
              " - Start (0)\\",\\"States\\":{\\"",
              Object {
                "Ref": "MyCanary1A94CAFA",
              },
              " - Start (0)\\":{\\"Next\\":\\"",
              Object {
                "Ref": "MyCanary1A94CAFA",
              },
              " - Wait 10s (0)\\",\\"Retry\\":[{\\"ErrorEquals\\":[\\"Lambda.ServiceException\\",\\"Lambda.AWSLambdaException\\",\\"Lambda.SdkClientException\\"],\\"IntervalSeconds\\":2,\\"MaxAttempts\\":6,\\"BackoffRate\\":2}],\\"Type\\":\\"Task\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke\\",\\"Parameters\\":{\\"FunctionName\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "stepFunctionOrchestratorstartCanaryF0F8EDC8",
                  "Arn",
                ],
              },
              "\\",\\"Payload\\":{\\"canaryName\\":\\"",
              Object {
                "Ref": "MyCanary1A94CAFA",
              },
              "\\"}}},\\"",
              Object {
                "Ref": "MyCanary1A94CAFA",
              },
              " - Wait 10s (0)\\":{\\"Type\\":\\"Wait\\",\\"Seconds\\":10,\\"Next\\":\\"",
              Object {
                "Ref": "MyCanary1A94CAFA",
              },
              " - Get Status (0)\\"},\\"",
              Object {
                "Ref": "MyCanary1A94CAFA",
              },
              " - Get Status (0)\\":{\\"Next\\":\\"",
              Object {
                "Ref": "MyCanary1A94CAFA",
              },
              " - Canary  Passed ? (0)\\",\\"Retry\\":[{\\"ErrorEquals\\":[\\"Lambda.ServiceException\\",\\"Lambda.AWSLambdaException\\",\\"Lambda.SdkClientException\\"],\\"IntervalSeconds\\":2,\\"MaxAttempts\\":6,\\"BackoffRate\\":2}],\\"Catch\\":[{\\"ErrorEquals\\":[\\"States.ALL\\"],\\"ResultPath\\":\\"$.State\\",\\"Next\\":\\"",
              Object {
                "Ref": "MyCanary1A94CAFA",
              },
              " - Silent fail (0)\\"}],\\"Type\\":\\"Task\\",\\"OutputPath\\":\\"$.Payload\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke\\",\\"Parameters\\":{\\"FunctionName\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "stepFunctionOrchestratorcheckCanary6F2AAE98",
                  "Arn",
                ],
              },
              "\\",\\"Payload\\":{\\"canaryName\\":\\"",
              Object {
                "Ref": "MyCanary1A94CAFA",
              },
              "\\"}}},\\"",
              Object {
                "Ref": "MyCanary1A94CAFA",
              },
              " - Canary  Passed ? (0)\\":{\\"Type\\":\\"Choice\\",\\"Choices\\":[{\\"Variable\\":\\"$.State\\",\\"StringEquals\\":\\"PASSED\\",\\"Next\\":\\"",
              Object {
                "Ref": "MyCanary1A94CAFA",
              },
              " - Canary 0 run Passed\\"}],\\"Default\\":\\"",
              Object {
                "Ref": "MyCanary1A94CAFA",
              },
              " - Wait 10s (0)\\"},\\"",
              Object {
                "Ref": "MyCanary1A94CAFA",
              },
              " - Canary 0 run Passed\\":{\\"Type\\":\\"Pass\\",\\"End\\":true},\\"",
              Object {
                "Ref": "MyCanary1A94CAFA",
              },
              " - Silent fail (0)\\":{\\"Type\\":\\"Pass\\",\\"Comment\\":\\"$.StateReason\\",\\"End\\":true}}}]},\\"checkResults\\":{\\"Type\\":\\"Map\\",\\"End\\":true,\\"Iterator\\":{\\"StartAt\\":\\"Double check result\\",\\"States\\":{\\"Double check result\\":{\\"Type\\":\\"Choice\\",\\"Choices\\":[{\\"Variable\\":\\"$.State\\",\\"StringEquals\\":\\"PASSED\\",\\"Next\\":\\"Pass\\"}],\\"Default\\":\\"Fail\\"},\\"Fail\\":{\\"Type\\":\\"Fail\\"},\\"Pass\\":{\\"Type\\":\\"Pass\\",\\"End\\":true}}}}},\\"TimeoutSeconds\\":900}",
            ],
          ],
        },
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "stepFunctionOrchestratorE2ETestsRunnerRole6F118857",
            "Arn",
          ],
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
    },
    "stepFunctionOrchestratorE2ETestsRunnerRole6F118857": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "states.",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ".amazonaws.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "stepFunctionOrchestratorE2ETestsRunnerRoleDefaultPolicy267BF1ED": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "stepFunctionOrchestratorstartCanaryF0F8EDC8",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "stepFunctionOrchestratorcheckCanary6F2AAE98",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "stepFunctionOrchestratorE2ETestsRunnerRoleDefaultPolicy267BF1ED",
        "Roles": Array [
          Object {
            "Ref": "stepFunctionOrchestratorE2ETestsRunnerRole6F118857",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "stepFunctionOrchestratorcheckCanary6F2AAE98": Object {
      "DependsOn": Array [
        "stepFunctionOrchestratorcheckCanaryServiceRoleDefaultPolicyC45E0816",
        "stepFunctionOrchestratorcheckCanaryServiceRole4593C5FC",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters49cd87fb06607303bf223aa06c1cef28acff95117850c08947a62d8257902d6dS3Bucket81F354A2",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters49cd87fb06607303bf223aa06c1cef28acff95117850c08947a62d8257902d6dS3VersionKey17FB176D",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters49cd87fb06607303bf223aa06c1cef28acff95117850c08947a62d8257902d6dS3VersionKey17FB176D",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "stepFunctionOrchestratorcheckCanaryServiceRole4593C5FC",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "stepFunctionOrchestratorcheckCanaryServiceRole4593C5FC": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "stepFunctionOrchestratorcheckCanaryServiceRoleDefaultPolicyC45E0816": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "synthetics:GetCanaryRuns",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "stepFunctionOrchestratorcheckCanaryServiceRoleDefaultPolicyC45E0816",
        "Roles": Array [
          Object {
            "Ref": "stepFunctionOrchestratorcheckCanaryServiceRole4593C5FC",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "stepFunctionOrchestratorstartCanaryF0F8EDC8": Object {
      "DependsOn": Array [
        "stepFunctionOrchestratorstartCanaryServiceRoleDefaultPolicyB1AB8CFB",
        "stepFunctionOrchestratorstartCanaryServiceRole34FE2D90",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersa092c9a65e948c08fdd58344d465f3fe147850fd89e1402f7354ce067360eb03S3Bucket7E06B8E8",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersa092c9a65e948c08fdd58344d465f3fe147850fd89e1402f7354ce067360eb03S3VersionKey8A44DBB6",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersa092c9a65e948c08fdd58344d465f3fe147850fd89e1402f7354ce067360eb03S3VersionKey8A44DBB6",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "stepFunctionOrchestratorstartCanaryServiceRole34FE2D90",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "stepFunctionOrchestratorstartCanaryServiceRole34FE2D90": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "stepFunctionOrchestratorstartCanaryServiceRoleDefaultPolicyB1AB8CFB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "synthetics:StartCanary",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "stepFunctionOrchestratorstartCanaryServiceRoleDefaultPolicyB1AB8CFB",
        "Roles": Array [
          Object {
            "Ref": "stepFunctionOrchestratorstartCanaryServiceRole34FE2D90",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
